### PR TITLE
GODRIVER-2779

### PR DESCRIPTION
**Pull request by qingyang-hu**
GODRIVER-2779
GODRIVER-1892

## Summary
1. Merge the memory improvement in PR #1199;
2. Add bson.UnmarshalValue function;
3. Use bson.UnmarshalValue to test the changes for memory improvement.

## Background & Motivation

Using a Marshaler/Unmarshaler seems to be a more comprehensive way to test the #1199.

### CODE REVIEW

File: bson/marshal_value_test.go [hunk 1]
```diff
@@ -7,100 +7,16 @@
...
-		{"timestamp", primitive.Timestamp{T: 1, I: 5}, bsontype.Timestamp, bsoncore.AppendTimestamp(nil, 1, 5)},
-		{"decimal128", decimal128, bsontype.Decimal128, bsoncore.AppendDecimal128(nil, decimal128)},
-		{"min key", primitive.MinKey{}, bsontype.MinKey, []byte{}},
-		{"max key", primitive.MaxKey{}, bsontype.MaxKey, []byte{}},
-		{"struct", structTest, bsontype.EmbeddedDocument, structCore},
-		{"interface", interfaceTest, bsontype.EmbeddedDocument, interfaceCore},
-		{"D", D{{"foo", 10}}, bsontype.EmbeddedDocument, structCore},
-		{"M", M{"foo": 10}, bsontype.EmbeddedDocument, structCore},
-		{"ValueMarshaler", marshalValueMarshaler{Foo: 10}, bsontype.Int32, bsoncore.AppendInt32(nil, 10)},
-	}
```
Comment (qingyang-hu) [inline bson/marshal_value_test.go:103]: Test cases are moved to "bson/unmarshal_value_test.go".

File: bson/marshalvalue_cases_test.go [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+}
+
+type marshalValueTestCase struct {
+	name     string
+	val      interface{}
+	bsontype bsontype.Type
+	bytes    []byte
+}
+
+func unmarshalValueTestCases(t *testing.T) []marshalValueTestCase {
```
Comment (prestonvasquez) [inline bson/marshalvalue_cases_test.go:65]: Why do we call this function "unmarshalValueTestCases" if it returns a slice of "marshalValueTestCase"? Maybe a better name would be "newMarshalValueTestCases." What are your thoughts?

File: bson/marshalvalue_cases_test.go [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+	oid := primitive.NewObjectID()
+	regex := primitive.Regex{Pattern: "pattern", Options: "imx"}
+	dbPointer := primitive.DBPointer{DB: "db", Pointer: primitive.NewObjectID()}
+	codeWithScope := primitive.CodeWithScope{Code: "code", Scope: D{{"a", "b"}}}
+	idx, scopeCore := bsoncore.AppendDocumentStart(nil)
+	scopeCore = bsoncore.AppendStringElement(scopeCore, "a", "b")
+	scopeCore, _ = bsoncore.AppendDocumentEnd(scopeCore, idx)
+	decimal128 := primitive.NewDecimal128(5, 10)
+	structTest := marshalValueStruct{Foo: 10}
+	structCore, err := Marshal(structTest)
```
Comment (prestonvasquez) [inline bson/marshalvalue_cases_test.go:77]: Wrapping this data in `var()` would align the "=" values and make this more readable.

File: bson/marshalvalue_cases_test.go [hunk 4]
```diff
@@ -0,0 +1,127 @@
+// Copyright (C) MongoDB, Inc. 2023-present.
```
Comment (prestonvasquez) [inline bson/marshalvalue_cases_test.go:1]: Should this file be "bson/marshalvalue_cases_test.go" or "bson/marshal_value_cases_test.go" to (sort of) name it after the "marshalValueTestCase" struct?

File: bson/marshal_value_test.go [hunk 5]
```diff
@@ -162,7 +78,20 @@ func TestMarshalValue(t *testing.T) {
...
 	gotValue := RawValue{Type: gotType, Value: gotBytes}
 	assert.Equal(t, expectedValue, gotValue, "value mismatch; expected %s, got %s", expectedValue, gotValue)
 }
+
+// benchmark covering GODRIVER-2779
+func BenchmarkSliceCodec_Marshal(b *testing.B) {
+	testStruct := unmarshalerNonPtrStruct{B: []byte(strings.Repeat("t", 4096))}
+
+	b.ResetTimer()
+	for n := 0; n < b.N; n++ {
```
Comment (prestonvasquez) [inline bson/marshal_value_test.go:91]: Should we parallelize this? I.e. something like

```go
b.RunParallel(func(pb *testing.PB) {
      for pb.Next() {
	      _, _, err := MarshalValueWithRegistry(DefaultRegistry, testStruct)
	      if err != nil {
		      b.Error(err)
	      }
      }
})
```

File: bson/marshal_value_test.go [hunk 6]
```diff
@@ -162,7 +78,20 @@ func TestMarshalValue(t *testing.T) {
...
 func compareMarshalValueResults(t *testing.T, tc marshalValueTestCase, gotType bsontype.Type, gotBytes []byte) {
 	t.Helper()
-	expectedValue := RawValue{Type: tc.expectedType, Value: tc.expectedBytes}
+	expectedValue := RawValue{Type: tc.bsontype, Value: tc.bytes}
 	gotValue := RawValue{Type: gotType, Value: gotBytes}
 	assert.Equal(t, expectedValue, gotValue, "value mismatch; expected %s, got %s", expectedValue, gotValue)
 }
+
+// benchmark covering GODRIVER-2779
+func BenchmarkSliceCodec_Marshal(b *testing.B) {
```
Comment (prestonvasquez) [inline bson/marshal_value_test.go:87]: What is the purpose of the "_" in the benchmark function name? Should it be removed?

File: bson/marshalvalue_cases_test.go [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+		{"max key", primitive.MaxKey{}, bsontype.MaxKey, []byte{}},
+		{"struct", structTest, bsontype.EmbeddedDocument, structCore},
+		{"D", D{{"foo", int32(10)}}, bsontype.EmbeddedDocument, structCore},
+		{"M", M{"foo": int32(10)}, bsontype.EmbeddedDocument, structCore},
+		{"ValueMarshaler", marshalValueMarshaler{Foo: 10}, bsontype.Int32, bsoncore.AppendInt32(nil, 10)},
+	}
+
+}
+
+func marshalValueTestCases(t *testing.T) []marshalValueTestCase {
```
Comment (prestonvasquez) [inline bson/marshalvalue_cases_test.go:108]: If the change "newMarshalValueTestCases" is accepted, then this could be something like "newMarshalValueTestCasesWithInterfaceCore".

File: bson/marshalvalue_cases_test.go [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+		Reader: marshalValueInterfaceInner{
+			Foo: 10,
+		},
+	}
+	interfaceCore, err := Marshal(interfaceTest)
+	assert.Nil(t, err, "Marshal error: %v", err)
+
+	marshalValueTestCases = append(
+		marshalValueTestCases,
+		marshalValueTestCase{"interface", interfaceTest, bsontype.EmbeddedDocument, interfaceCore},
```
Comment (prestonvasquez) [inline bson/marshalvalue_cases_test.go:123]: What is the purpose of adding this here instead of with the other "marshalValueTestCase" in "unmarshalValueTestCases" ?

File: bson/unmarshal.go [hunk 9]
```diff
@@ -61,6 +61,21 @@ func UnmarshalWithContext(dc bsoncodec.DecodeContext, data []byte, val interface
 	return unmarshalFromReader(dc, vr, val)
 }
 
+// UnmarshalValue parses the BSON value of type t with bson.DefaultRegistry and
+// stores the result in the value pointed to by val. If val is nil or not a pointer,
+// UnmarshalValue returns an error.
+func UnmarshalValue(t bsontype.Type, data []byte, val interface{}) error {
```
Comment (prestonvasquez) [inline bson/unmarshal.go:67]: The addition of "UnmarshalValue" is scheduled as a 2.0 improvement in [GODRIVER-1892](https://jira.mongodb.org/browse/GODRIVER-1892). I don't think we should add it here unless GODRIVER-2779 is marked a 2.0 and 1892 a duplicate.

For the sake of this ticket, perhaps it would be best to make these unexported test helpers. We could mark them as "TODO: GODRIVER-1892 remove with the introduction of stable API UnmarshalValue".

@matthewdale Do you have thoughts on this?

File: bson/unmarshal.go [hunk 10]
```diff
@@ -61,6 +61,21 @@ func UnmarshalWithContext(dc bsoncodec.DecodeContext, data []byte, val interface
...
+	return UnmarshalValueWithRegistry(DefaultRegistry, t, data, val)
+}
+
+// UnmarshalValueWithRegistry parses the BSON value of type t with registry r and
+// stores the result in the value pointed to by val. If val is nil or not a pointer,
+// UnmarshalValue returns an error.
+func UnmarshalValueWithRegistry(r *bsoncodec.Registry, t bsontype.Type, data []byte, val interface{}) error {
```
Comment (prestonvasquez) [inline bson/unmarshal.go:74]: See comment on UnmarshalValue.

**Review by prestonvasquez — CHANGES_REQUESTED** [review]
The changes look good so far, I am concerned about the addition of UnmarshalValue.

File: bson/unmarshal.go (hunk 9 above)
Comment (matthewdale) [reply bson/unmarshal.go:67]: GODRIVER-1892 is part of the Go Driver 2.0 preparation epic, which is a collection of non-breaking changes we want to make in Go Driver 1.x that makes transition to Go Driver 2.0 easier. We can make this additive API change in any Go Driver 1.x release, which will remain part of the Go Driver 2.0 API.

TLDR: We should keep this change because it is strictly additive and should be part of the Go Driver 1.x and 2.x API.

File: bson/unmarshal.go (hunk 10 above)
Comment (matthewdale) [reply bson/unmarshal.go:74]: We should remove this API because we are deprecating all `*WithRegistry` marshaling functions with plans to remove them in Go Driver 2.0. Instead, we should provide `Encoder.MarshalValue` and `Decoder.UnmarshalValue` methods. We can add those in this PR or in a subsequent PR.

Comment (qingyang-hu) [reply bson/unmarshal.go:74]: What is the plan for the `*WithRegistry`? How do you think keeping the `UnmarshalValueWithRegistry` in this PR to keep the code consistent before a subsequent `*WithRegistry` PR?

Comment (matthewdale) [reply bson/unmarshal.go:74]: Our plan is to deprecate all `*WithRegistry` functions and recommend that users use a `bson.Encoder` or `bson.Decoder` instead if they need to use a custom `Registry` (see [GODRIVER-2711](https://jira.mongodb.org/browse/GODRIVER-2711)). However, for symmetry of the current API, we can keep this with the understanding that it will be deprecated soon after it's added.

File: bson/unmarshal_value_test.go [hunk 11]
```diff
@@ -0,0 +1,108 @@
...
+	"testing"
+
+	"go.mongodb.org/mongo-driver/bson/bsoncodec"
+	"go.mongodb.org/mongo-driver/bson/bsontype"
+	"go.mongodb.org/mongo-driver/internal/assert"
+	"go.mongodb.org/mongo-driver/x/bsonx/bsoncore"
+)
+
+func TestUnmarshalValue(t *testing.T) {
+	unmarshalValueTestCases := newMarshalValueTestCases(t)
```
Comment (prestonvasquez) [inline bson/unmarshal_value_test.go:21]:
```suggestion
func TestUnmarshalValue(t *testing.T) {
	t.Parallel()

	unmarshalValueTestCases := newMarshalValueTestCases(t)
```

File: bson/unmarshal_value_test.go [hunk 12]
```diff
@@ -0,0 +1,108 @@
...
+
+	t.Run("UnmarshalValue", func(t *testing.T) {
+		for _, tc := range unmarshalValueTestCases {
```
Comment (prestonvasquez) [inline bson/unmarshal_value_test.go:24]:
```suggestion
	t.Run("UnmarshalValue", func(t *testing.T) {
		t.Parallel()

		for _, tc := range unmarshalValueTestCases {
			tc := tc
```

File: bson/unmarshal_value_test.go [hunk 13]
```diff
@@ -0,0 +1,108 @@
...
+			t.Run(tc.name, func(t *testing.T) {
+				gotValue := reflect.New(reflect.TypeOf(tc.val))
```
Comment (prestonvasquez) [inline bson/unmarshal_value_test.go:26]:
```suggestion
			t.Run(tc.name, func(t *testing.T) {
				t.Parallel()

				gotValue := reflect.New(reflect.TypeOf(tc.val))
```

File: bson/unmarshal_value_test.go [hunk 14]
```diff
@@ -0,0 +1,108 @@
...
+				err := UnmarshalValue(tc.bsontype, tc.bytes, gotValue.Interface())
+				assert.Nil(t, err, "UnmarshalValueWithRegistry error: %v", err)
+				assert.Equal(t, tc.val, gotValue.Elem().Interface(), "value mismatch; expected %s, got %s", tc.val, gotValue.Elem())
+			})
+		}
+	})
+	t.Run("UnmarshalValueWithRegistry with DefaultRegistry", func(t *testing.T) {
+		for _, tc := range unmarshalValueTestCases {
```
Comment (prestonvasquez) [inline bson/unmarshal_value_test.go:34]:
```suggestion
	t.Run("UnmarshalValueWithRegistry with DefaultRegistry", func(t *testing.T) {
		t.Parallel()

		for _, tc := range unmarshalValueTestCases {
			tc := tc
```

File: bson/unmarshal_value_test.go [hunk 15]
```diff
@@ -0,0 +1,108 @@
...
+			t.Run(tc.name, func(t *testing.T) {
+				gotValue := reflect.New(reflect.TypeOf(tc.val))
```
Comment (prestonvasquez) [inline bson/unmarshal_value_test.go:36]:
```suggestion
			t.Run(tc.name, func(t *testing.T) {
				t.Parallel()

				gotValue := reflect.New(reflect.TypeOf(tc.val))
```

File: bson/unmarshal_value_test.go [hunk 16]
```diff
@@ -0,0 +1,108 @@
...
+				err := UnmarshalValueWithRegistry(DefaultRegistry, tc.bsontype, tc.bytes, gotValue.Interface())
+				assert.Nil(t, err, "UnmarshalValueWithRegistry error: %v", err)
+				assert.Equal(t, tc.val, gotValue.Elem().Interface(), "value mismatch; expected %s, got %s", tc.val, gotValue.Elem())
+			})
+		}
+	})
+	// tests covering GODRIVER-2779
+	t.Run("UnmarshalValueWithRegistry with custom registry", func(t *testing.T) {
+		testCases := []struct {
```
Comment (prestonvasquez) [inline bson/unmarshal_value_test.go:45]:
```suggestion
	t.Run("UnmarshalValueWithRegistry with custom registry", func(t *testing.T) {
		t.Parallel()

		testCases := []struct {
```

File: bson/unmarshal_value_test.go [hunk 17]
```diff
@@ -0,0 +1,108 @@
...
+				name:     "SliceCodec string",
+				val:      []byte("hello world"),
+				bsontype: bsontype.String,
+				bytes:    bsoncore.AppendString(nil, "hello world"),
+			},
+		}
+		rb := NewRegistryBuilder().RegisterTypeDecoder(reflect.TypeOf([]byte{}), bsoncodec.NewSliceCodec()).Build()
+		for _, tc := range testCases {
+			t.Run(tc.name, func(t *testing.T) {
+				gotValue := reflect.New(reflect.TypeOf(tc.val))
```
Comment (prestonvasquez) [inline bson/unmarshal_value_test.go:67]:
```suggestion
		for _, tc := range testCases {
			tc := tc 

			t.Run(tc.name, func(t *testing.T) {
				t.Parallel()

				gotValue := reflect.New(reflect.TypeOf(tc.val))
```

File: bson/unmarshal_value_test.go [hunk 18]
```diff
@@ -0,0 +1,108 @@
...
+			bytes:    bsoncore.AppendString(nil, strings.Repeat("t", 4096)),
+		},
+	}
+	rb := NewRegistryBuilder().RegisterTypeDecoder(reflect.TypeOf([]byte{}), bsoncodec.NewSliceCodec()).Build()
+	for _, bm := range benchmarks {
+		b.Run(bm.name, func(b *testing.B) {
+			b.RunParallel(func(pb *testing.PB) {
+				for pb.Next() {
+					var gotValue []byte
+					err := UnmarshalValueWithRegistry(rb, bm.bsontype, bm.bytes, &gotValue)
```
Comment (prestonvasquez) [inline bson/unmarshal_value_test.go:100]: Optional, define assignment object inline since we don't use gotValue.

```suggestion
					err := UnmarshalValueWithRegistry(rb, bm.bsontype, bm.bytes, &[]byte{})
```

**Review by prestonvasquez — APPROVED** [review]
LGTM

**Review by matthewdale — APPROVED** [review]
One requested docs change, but looks good! 👍
